pyro/contrib/tracking/hashing.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import absolute_import, division, print_function
+
+import itertools
+from collections import defaultdict
+from numbers import Number
+
+
+class LSH(object):
+    """
+    Implements locality-sensitive hashing. Allows to efficiently find neighbours
+    of a point. Provides 2 guarantees:
+    - Difference between coordinates of points not returned by :meth:`nearby`
+    and input point is larger than `radius`.
+    - Difference between coordinates of points returned by :meth:`nearby` and
+    input point is smaller than 2`radius`.
+    :param float radius: scaling parameter used in hash function. Determines
+    the size of the neighbourhood.
+
+    Example::
+        radius = 1
+        lsh = LSH(radius)
+        a = torch.tensor([-0.49, -0.49]) # hash(a)=(0,0)
+        b = torch.tensor([1.0, 1.0]) # hash(b)=(1,1)
+        c = torch.tensor([2.49, 2.49]) # hash(c)=(2,2)
+        lsh.add('a', a)
+        lsh.add('b', b)
+        lsh.add('c', c)
+        lsh.nearby('a') # {'b'}
+        lsh.nearby('b') # {'a','c'}
+        lsh.remove('b')
+        lsh.nearby('a') # {}
+    """
+    def __init__(self, radius):
+        assert radius > 0 if isinstance(radius, Number) else (radius > 0).all(), \
+            "radius must be greater than 0, given: {}".format(radius)
+        self._radius = radius
+        self._hash_to_key = defaultdict(set)
+        self._key_to_hash = {}
+
+    def _hash(self, point):
+        coords = (point.detach().cpu() / self._radius).round()
+        return tuple(map(int, coords))
+
+    def add(self, key, point):
+        """
+        Adds (`key`, `hash(point)`) pair to the hash
+        :param key: key used identify point
+        :param point: data, should be hashable
+        """
+        _hash = self._hash(point)
+        if key in self._key_to_hash.keys():
```

Comment:
Remove the unnecessary `.keys()`. In Python2 this creates an entire copy of keys list each call.